File: 2022/scala/aoc2022/src/main/scala/aoc/Day6.scala
package aoc

import scala.annotation.tailrec

object Day6 extends Day {
  override def part1(s: String): String =
    marker(s, 0, 4).toString

  override def part2(s: String): String =
    marker(s, 0, 14).toString

  @tailrec
  def marker(s: String, i: Int, len: Int): Int =
    if (s.slice(i, i+len).distinct.length == len) i+len
    else marker(s, i+1, len)

}

File: 2022/rust/aoc2022/src/day6.rs
use crate::day;
use itertools::Itertools;
pub struct Day6 {}

impl Day6 {}

fn marker(cs: &[char], i: usize, len: usize) -> usize {
    if cs[i..i + len]
        .into_iter()
        .copied()
        .unique()
        .collect::<Vec<char>>()
        .len()
        == len
    {
        i + len
    } else {
        marker(cs, i + 1, len)
    }
}

impl day::Solution for Day6 {
    fn part1(s: &String) -> String {
        format!("{}", marker(&s.chars().collect::<Vec<char>>(), 0, 4))
    }

    fn part2(s: &String) -> String {
        format!("{}", marker(&s.chars().collect::<Vec<char>>(), 0, 14))
    }
}
